presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/ReentrantBoundedExecutor.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.hive.metastore.cache;
+
+import io.airlift.concurrent.BoundedExecutor;
+
+import java.util.concurrent.Executor;
+
+import static java.util.Objects.requireNonNull;
+
+public class ReentrantBoundedExecutor
+        implements Executor
+{
+    private final ThreadLocal<Boolean> executorThreadMarkers = ThreadLocal.withInitial(() -> false);
+    private final Executor boundedExecutor;
+    private final Executor coreExecutor;
+
+    public ReentrantBoundedExecutor(Executor coreExecutor, int maxThreads)
+    {
+        this.boundedExecutor = new BoundedExecutor(requireNonNull(coreExecutor, "coreExecutor is null"), maxThreads);
+        this.coreExecutor = coreExecutor;
+    }
+
+    @Override
+    public void execute(Runnable task)
+    {
+        if (executorThreadMarkers.get()) {
+            coreExecutor.execute(task);
+        }
+        else {
+            // remove the initial marker value initialized by java.lang.ThreadLocal.get
+            executorThreadMarkers.remove();
+        }
```

Comment:
Remove the `else` block.
If you really want reset state, use no initial value (and `Object`, not `Boolean`).
If you want to use `Boolean`, then initial value is appropriate and you shouldn't reset `false` to unset.